feat: Simplify Mover interface

migmartri opened the pull request:
Split from https://github.com/vmware-tanzu/asset-relocation-tool-for-kubernetes/pull/37

Changes the mover constructor signature to not to require the caller to import and load neither the Helm Chart nor the imagehints file. A path to those will suffice now. 

Signed-off-by: Miguel Martinez Trivino <mtrivino@vmware.com>

josvazg commented on pkg/mover/chart.go:
```diff
@@ -4,18 +4,28 @@
+)
+
+var (
+	ErrImageHintsMissing  = errors.New("no image hints provided")
+	ErrOCIRewritesMissing = errors.New("at least one rewrite rule is required")
```
These usually require comments, isn't the editor requesting them?

migmartri replied:
Right, my editor does not request them because my default linter is `golangci-lint`, which does not include by default golint (or revive) until this pr
https://github.com/vmware-tanzu/asset-relocation-tool-for-kubernetes/pull/48 is merged, 

If we merge that PR, golangci-lint configured editors will complain as well as our CI would have complained.

migmartri replied:
Actually I was wrong. The issue is that golangci-lint has actually disabled that rule 

I have created a task to enable it (if we find it appropriate) https://github.com/vmware-tanzu/asset-relocation-tool-for-kubernetes/issues/50

josvazg commented on pkg/mover/chart.go:
```diff
@@ -4,18 +4,28 @@
 )
 
 // Number of retries for pull/push operations
-const DefaultRetries = 3
+const (
+	DefaultRetries        = 3
```
The comment must move with the definition of the constant, right?

josvazg reviewed (approved):
I like this change, it reduces our public interface (contract surface) while keeping the functionality. This is something I was trying to do as part of moving code to a library, and this change improves on that.

I only have some questions/issues about comments, it would be great if those could be fixed before merging.

josvazg commented:
Our full public API right now after this change. Apart fro the misplaced comments LGTM:
~~~bash
$ go doc -all pkg/mover/chart.go 
~~~
~~~go
package mover // import "github.com/vmware-tanzu/asset-relocation-tool-for-kubernetes/v2/pkg/mover"


CONSTANTS

const (
	DefaultRetries        = 3
	EmbeddedHintsFilename = ".relok8s-images.yaml"
)
    Number of retries for pull/push operations


VARIABLES

var (
	ErrImageHintsMissing  = errors.New("no image hints provided")
	ErrOCIRewritesMissing = errors.New("at least one rewrite rule is required")
)

TYPES

type ChartMover struct {

	// Extracted metadata from the provided Helm Chart
	ChartName    string
	ChartVersion string

	// Has unexported fields.
}

func NewChartMover(chartPath string, imageHintsFile string, rules *RewriteRules, opts ...Option) (*ChartMover, error)
    NewChartMover creates a ChartMover to relocate a chart following the given
    imagePatters and rules.

func (cm *ChartMover) Move(toChartFilename string) error
    Move executes the chart move image and chart changes. The result chart is
    saved as toChartFilename.

func (cm *ChartMover) Print()
    Print dumps the chart mover changes to the mover logger

func (cm *ChartMover) WithRetries(retries uint) *ChartMover
    WithRetries customizes the mover push retries

type Logger interface {
	Printf(format string, i ...interface{})
	Println(i ...interface{})
}

type Option func(*ChartMover)

func WithLogger(l Logger) Option
    WithLogger sets a custom Logger interface

func WithRetries(retries uint) Option
    WithRetries defines how many times to retry the push operation

type RewriteRules struct {
	Registry         string
	RepositoryPrefix string
}
~~~

BTW we should probably check this doc from time to time to sure it makes sense to end users. Or we might even want to automate its generation somewhere so we can look at it easily.

petewall commented on test/chart_move_feature_test.go:
```diff
@@ -71,7 +71,7 @@ var _ = Describe("relok8s chart move command", func() {
 
 	Scenario("missing image patterns file", func() {
-		steps.When("running relok8s chart move fixtures/wordpress-11.0.4.tgz")
+		steps.When("running relok8s chart move fixtures/wordpress-11.0.4.tgz --repo-prefix cyberdyne-corp")
```
nice

migmartri commented:
I will spend some time making sure that the documentation is correct.

Thanks for pointing this out, it's very important and I did a messy job there.

migmartri commented:
> I like this change, it reduces our public interface (contract surface) while keeping the functionality. This is something I was trying to do as part of moving code to a library, and this change improves on that.
> 
Glad to hear, yes, that was the goal and it was perfect to work on top of your foundation :)

> I only have some questions/issues about comments, it would be great if those could be fixed before merging.

Yeah, I'll fix them and make sure the generated documentation quality is good.

Thanks!

migmartri commented:
I have updated the documentation now as well as added an example

![2021-08-23_12-37](https://user-images.githubusercontent.com/24523/130508833-50310e05-270b-40ee-979f-1486004c2c3d.png)


Also, while documenting it, I realized that my approach of exposing metadata could be confusing, since it could be mistaken as a way to provide Helm Chart input. I have updated the approach to now instead have a standalone getMetadata method.
